libs/core/algorithms/tests/unit/algorithms/uninitialized_relocate_backward.cpp

```diff
@@ -0,0 +1,308 @@
+
+        HPX_TEST(non_trivially_relocatable_struct::count == 0);
+        HPX_TEST(non_trivially_relocatable_struct::move_count == 0);
+        HPX_TEST(non_trivially_relocatable_struct::dtor_count == 0);
+
+        for (int i = 0; i < N; i++)
+        {
+            hpx::construct_at(ptr1 + i, 1234);
+            hpx::construct_at(ptr2 + i, 5678);
+        }
+
+        // N objects constructed
+        HPX_TEST(non_trivially_relocatable_struct::count == 2 * N);
+
+        // relocate them to ptr2
+        uninitialized_relocate_backward(ptr1, ptr1 + M, ptr2 + N);
+
+        // M move constructors were called and M destructors
+        HPX_TEST(non_trivially_relocatable_struct::move_count == M);
+        HPX_TEST(non_trivially_relocatable_struct::dtor_count == M);
+
+        for (int i = M; i < N; i++)
+        {
+            HPX_TEST(ptr1[i].data == 1234);
+        }
+
+        for (int i = 0; i < N - M; i++)
+        {
+            HPX_TEST(ptr2[i].data == 5678);
+        }
+
+        for (int i = N - M; i < N; i++)
+        {
+            HPX_TEST(ptr2[i].data == 1234);
+        }
+
+        std::destroy(ptr1 + M, ptr1 + N);
+        std::destroy(ptr2, ptr2 + N);
+
+        std::free(mem1);
+        std::free(mem2);
+    }
+    {
+        void* mem1 =
+            std::malloc(N * sizeof(non_trivially_relocatable_struct_throwing));
+        void* mem2 =
+            std::malloc(N * sizeof(non_trivially_relocatable_struct_throwing));
+
+        HPX_TEST(mem1 && mem2);
+
+        non_trivially_relocatable_struct_throwing* ptr1 =
+            static_cast<non_trivially_relocatable_struct_throwing*>(mem1);
+        non_trivially_relocatable_struct_throwing* ptr2 =
+            static_cast<non_trivially_relocatable_struct_throwing*>(mem2);
+
+        HPX_TEST(non_trivially_relocatable_struct_throwing::count == 0);
+        HPX_TEST(non_trivially_relocatable_struct_throwing::move_count == 0);
+        HPX_TEST(non_trivially_relocatable_struct_throwing::dtor_count == 0);
+
+        for (int i = 0; i < N; i++)
+        {
+            hpx::construct_at(ptr1 + i, 1234);
+            hpx::construct_at(ptr2 + i, 5678);
+        }
+
+        // N objects constructed
+        HPX_TEST(non_trivially_relocatable_struct_throwing::count == 2 * N);
+
+        // relocate them to ptr2
+        try
+        {
+            uninitialized_relocate_backward(ptr1, ptr1 + M, ptr2 + N);
+            HPX_UNREACHABLE;    // should have thrown
+        }
+        catch (...)
+        {
+        }
+
+        // K move constructors were called, and then the last one throws
+        HPX_TEST(non_trivially_relocatable_struct_throwing::move_count == K);
+
+        // K - 1 destructors were called to balance out the move constructors
+        // (- 1 because the last move constructor throws)
+        // and then M + 1 destructors were called: K on the old range and
+        // M - (K - 1) = M - K + 1 on the new range
```

Comment:
FWIW, I found this explanation hard to grasp. I would prefer to say that there are M+K-1 objects destroyed because we successfully relocated K-1 objects before the failure (so that's K-1 new objects we need to destroy), plus when we fail we destroy the entire source range (so that's M objects destroyed _at some point_, although of course some of those destructions have already happened as part of the K-1 successful relocations).
...But you got M+K, not M+K-1. Maybe I'm missing one somewhere?